packages/webdriverio/src/types.ts

```diff
@@ -11,9 +11,78 @@ import type DevtoolsInterception from './utils/interception/devtools'
+type $BrowserCommands = typeof BrowserCommands
+type $ElementCommands = typeof ElementCommands
+
+type ElementQueryCommands = '$' | 'custom$' | 'shadow$' | 'react$'
+type ElementsQueryCommands = '$$' | 'custom$$' | 'shadow$$' | 'react$$'
+type ChainablePrototype = {
+    [K in ElementQueryCommands]: (...args: Parameters<$ElementCommands[K]>) => ChainablePromiseElement<ReturnType<$ElementCommands[K]>>
+} & {
+    [K in ElementsQueryCommands]: (...args: Parameters<$ElementCommands[K]>) => ChainablePromiseArray<WebdriverIO.Element>
+}
+
+type AsyncElementProto = {
+    [K in keyof Omit<$ElementCommands, keyof ChainablePrototype>]: OmitThisParameter<$ElementCommands[K]>
+} & ChainablePrototype
+
+interface ChainablePromiseElement<T> extends AsyncElementProto, Promise<T> {
+    /**
+     * WebDriver element reference
+     */
+    elementId: Promise<string>
+    /**
+     * parent of the element if fetched via `$(parent).$(child)`
+     */
+    parent: Promise<WebdriverIO.Element | WebdriverIO.Browser>
+    /**
+     * selector used to fetch this element, can be
+     * - undefined if element was created via `$({ 'element-6066-11e4-a52e-4f735466cecf': 'ELEMENT-1' })`
+     * - a string if `findElement` was used and a reference was found
+     * - or a functin if element was found via e.g. `$(() => document.body)`
+     */
+    selector: Promise<Selector>
+    /**
+     * Error message in case element fetch was not successful
+     */
+    error?: Promise<Error>
+    /**
+     * index of the element if fetched with `$$`
+     */
+    index?: Promise<number>
+}
+interface ChainablePromiseArray<T> extends Promise<T> {
+    /**
+     * Amount of element fetched.
+     */
+    length: Promise<number>
+
+    /**
+     * Unwrap the nth element of the element list.
+     */
+    get: (i: number) => T
+    forEach: (iterator: (elem: T) => void) => ChainablePromiseArray<T>
+    forEachSeries: (iterator: (elem: T) => void) => ChainablePromiseArray<T>
+    map: (iterator: (elem: T) => any) => ChainablePromiseArray<T>
+    mapSeries: (iterator: (elem: T) => any) => ChainablePromiseArray<T>
+    find: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    findSeries: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    findIndex: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    findIndexSeries: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    some: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    someSeries: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    every: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    everySeries: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    filter: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    filterSeries: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    reduce: (iterator: (initialValue: any, elem: T) => any, initialValue: any) => any
```

Comment:
Origina Array's `reduce` method definition looks like this 
```
reduce<U>(callbackfn: (previousValue: U, currentValue: T, currentIndex: number, array: T[]) => U, initialValue: U): U;
```
Is it possible to adjust our wrapper accordingly?